README.md
# Stopwatch

Show the time that was necessary to render your page and the amount of queries that were executed.

## Installation

Just add the following to your __development__ group and you are good to go:

``` ruby
gem 'stopwatch'
```

Every page will now display the duration in the upper right corner and the
amount of (non-cached) queries.

![Stopwatch in action](https://github.com/bittersweet/stopwatch/raw/master/stopwatch.png)

## Inspiration

I was inspired by an article on the [Coding Horror
blog](http://www.codinghorror.com/blog/2011/06/performance-is-a-feature.html)
where they talk about how adding a small box that displays page speed motivated
them to not forget about the performance of their pages.

Some excerpts:

> The simple act of putting a render time in the upper right hand corner of
> every page we serve forced us to fix all our performance regressions and
> omissions.

> In fact, with the render time appearing on every page for everyone on the dev
> team, performance became a point of pride. We had so many places where we had
> just gotten a little sloppy or missed some tiny thing that slowed a page down
> inordinately.

I could especially identify with the last one and I wanted to quickly get this
gem out there to see if I could recreate that same effect.

## Caveats

* It removes etags so we can display it fresh every page load, otherwise the
response would get cached by the browser because Rails would return a 304 Not
Modified and you wouldn't see any changes on page refresh.

## Todo

* Basic configuration

## Thanks

[Paul Engel](https://github.com/archan937) For the name!

## License

Copyright (C) 2011 by Mark Mulder

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.


CHANGELOG
v1.0.0
  * Handles UTF-8 characters in body, thanks @erikj for #2 !

v0.0.4
  * Only inserts code if the content-type is text/html or application/xhtml+xml

v0.0.3
  * Displays every partial that was rendered and the amount of queries it used

v0.0.2
  * Show total amount of queries

v0.0.1
  * First release, displays duration of rendering the action


spec/stopwatch_spec.rb
# encoding: utf-8
#
require_relative './../lib/stopwatch'
require_relative './../lib/load_speed'

require 'rack/lint'
require "rack/mock"

describe Rack::LoadSpeed do
  # Copied from https://github.com/rails/rails/blob/3e48bfd49730331d8f09d4d96bb9a58959017a9c/activesupport/lib/active_support/notifications/instrumenter.rb#L78
  # so we don't have to depend on AS
  class Event
    attr_reader :name, :time, :transaction_id, :payload, :children
    attr_accessor :end

    def initialize(name, start, ending, transaction_id, payload)
      @name           = name
      @payload        = payload.dup
      @time           = start
      @transaction_id = transaction_id
      @end            = ending
      @children       = []
      @duration       = nil
    end

    def duration
      @duration ||= 1000.0 * (self.end - time)
    end
  end

  it 'inserts performance data' do
    app = lambda { |env| [200, {'Content-Type' => 'text/html'}, ["<html><body>test</body></html>"]]}
    event = Event.new('test', 1, 2, '123', { identifier: "app/views/tests/index.html.erb" })
    Stopwatch::Log.reset_query_count
    Stopwatch::Log.reset_sub_query_count
    Stopwatch::Log.reset_events

    Stopwatch::Log.events << Stopwatch::Event.new(event)
    Stopwatch::Log.query_count = 12
    Stopwatch::Log.event = event

    response = Rack::LoadSpeed.new(app).call({})

    body = response[2][0]
    expect(body).to include "performance_code"
    expect(body).to include "<strong>1000</strong> ms"
    expect(body).to include "<strong>12</strong> queries"
  end

  it 'does nothing for JS' do
    app = lambda { |env| [200, {'Content-Type' => 'application/json'}, ["{'id': '1'}"] ] }
    event = Event.new('test', 1, 2, '123', {})
    response = Rack::LoadSpeed.new(app).call({})

    body = response[2][0]
    expect(body).to eq "{'id': '1'}"
  end

  it 'check content-length via Rack::Lint' do
    app = lambda { |env| [200, {'Content-Type' => 'text/html'}, ["<html><body>snowman ☃</body></html>"]]}
    loadspeed = Rack::LoadSpeed.new(app)
    request = Rack::MockRequest.new(Rack::Lint.new(loadspeed))
    response = request.get('/')
    expect(response.status).to eq 200
  end
end


spec/dummy/config/routes.rb
Dummy::Application.routes.draw do
  root :to => "welcome#index"
  match '/javascript_test', :to => "welcome#javascript_test"
end


spec/dummy/config/application.rb
require File.expand_path('../boot', __FILE__)

require "active_model/railtie"
require "active_record/railtie"
require "action_controller/railtie"
require "action_view/railtie"
require "action_mailer/railtie"

Bundler.require
require "stopwatch"

module Dummy
  class Application < Rails::Application
    config.middleware.use "Rack::LoadSpeed"

    # Settings in config/environments/* take precedence over those specified here.
    # Application configuration should go into files in config/initializers
    # -- all .rb files in that directory are automatically loaded.

    # Custom directories with classes and modules you want to be autoloadable.
    # config.autoload_paths += %W(#{config.root}/extras)

    # Only load the plugins named here, in the order given (default is alphabetical).
    # :all can be used as a placeholder for all plugins not explicitly named.
    # config.plugins = [ :exception_notification, :ssl_requirement, :all ]

    # Activate observers that should always be running.
    # config.active_record.observers = :cacher, :garbage_collector, :forum_observer

    # Set Time.zone default to the specified zone and make Active Record auto-convert to this zone.
    # Run "rake -D time" for a list of tasks for finding time zone names. Default is UTC.
    # config.time_zone = 'Central Time (US & Canada)'

    # The default locale is :en and all translations from config/locales/*.rb,yml are auto loaded.
    # config.i18n.load_path += Dir[Rails.root.join('my', 'locales', '*.{rb,yml}').to_s]
    # config.i18n.default_locale = :de

    # JavaScript files you want as :defaults (application.js is always included).
    # config.action_view.javascript_expansions[:defaults] = %w(jquery rails)

    # Configure the default encoding used in templates for Ruby 1.9.
    config.encoding = "utf-8"

    # Configure sensitive parameters which will be filtered from the log file.
    config.filter_parameters += [:password]
  end
end


lib/stopwatch/log.rb
module Stopwatch
  class Log
    @@query_count = 0
    @@sub_query_count = 0
    @@events = []

    def self.event
      @@event
    end
    def self.event=(event)
      @@event = event
    end

    def self.events
      @@events
    end
    def self.events=(events)
      @@events = events
    end

    def self.query_count
      @@query_count
    end
    def self.query_count=(query_count)
      @@query_count = query_count
    end

    def self.sub_query_count
      @@sub_query_count
    end

    def self.reset_query_count
      @@query_count = 0
    end

    def self.reset_sub_query_count
      @@sub_query_count = 0
    end

    def self.increment_query_count
      @@query_count += 1
    end

    def self.increment_sub_query_count
      @@sub_query_count += 1
    end

    def self.reset_events
      @@events = []
    end
  end
end


lib/stopwatch/railtie.rb
module Stopwatch
  class Railtie < Rails::Railtie
    initializer "newplugin.initialize" do |app|
      app.config.middleware.use "Rack::LoadSpeed"

      # Start processing
      ActiveSupport::Notifications.subscribe "start_processing.action_controller" do |*args|
        Stopwatch::Log.reset_query_count
        Stopwatch::Log.reset_sub_query_count
        Stopwatch::Log.reset_events
      end

      # Every query
      ActiveSupport::Notifications.subscribe "sql.active_record" do |name, start, finish, id, payload|
        if payload[:name] != "CACHE"
          Stopwatch::Log.increment_query_count
          Stopwatch::Log.increment_sub_query_count
        end
      end

      # Every partial render
      ActiveSupport::Notifications.subscribe(/render/) do |name, start, finish, id, payload|
        event = ActiveSupport::Notifications::Event.new(name, start, finish, id, payload)
        stopwatch_event = Stopwatch::Event.new(event)
        stopwatch_event.query_count = Stopwatch::Log.sub_query_count
        Stopwatch::Log.events << stopwatch_event
        Stopwatch::Log.reset_sub_query_count
      end

      # End of processing
      ActiveSupport::Notifications.subscribe "process_action.action_controller" do |*args|
        Stopwatch::Log.event = ActiveSupport::Notifications::Event.new(*args)
      end
    end
  end
end


lib/load_speed.rb
module Rack
  class LoadSpeed
    def initialize(app)
      @app = app
    end

    def call(env)
      env.delete("HTTP_IF_NONE_MATCH")
      status, headers, response = @app.call(env)

      if status == 200 && headers["Content-Type"] =~ /text\/html|application\/xhtml\+xml/
        body = ""
        response.each {|part| body << part}
        index = body.rindex("</body>")
        if index
          body.insert(index, performance_code)
          #
          # handle body w/ UTF8 characters for Ruby 1.9+ to avoid Rack::Lint::LintError
          #
          headers["Content-Length"] = (body.respond_to?(:bytesize) ? body.bytesize : body.length).to_s
          response = [body]
        end
      end

      [status, headers, response]
    end

  protected

    def performance_code
      events = "<table id='performance_table'><tr><td></td><td>duration (ms)</td><td>queries</td></tr>"
      Stopwatch::Log.events.each do |event|
        events << "<tr><td>#{event.template}</td><td>#{event.duration}</td><td>#{event.query_count}</td></tr>"
      end
      event = Stopwatch::Log.event
      events << "<tr><td>#{event.payload[:path]}</td><td>#{event.duration}</td><td>#{Stopwatch::Log.query_count}</td></tr>"
      events << "</table>"

      html = <<-EOF
<style>
  #performance_code {
    z-index: 1000;
    position: absolute;
    top: 0;
    right: 0;
    height: 25px;
    width: 140px;
    overflow: hidden;
    background-color: #DE7A93;
    color: white;
    padding: 0 10px 0 10px;
    line-height: 20pt;
    font-family: "menlo";
    font-size: 10pt;
    text-align: right;
  }

  #performance_code:hover {
    height: auto;
    width: auto;
    padding-bottom: 10px;
  }

  table#performance_table {
  }

  table#performance_table td {
    padding-right: 15px;
  }
</style>
<div id="performance_code">
  <strong>#{Stopwatch::Log.event.duration.to_i}</strong> ms
  <strong>#{Stopwatch::Log.query_count}</strong> queries
  #{events}
</div>
EOF
      html
    end
  end
end
